File: sdk/typescript/src/webhook.ts
import { createHmac, timingSafeEqual } from 'crypto';

/**
 * Verify the X-Dactyl-Signature HMAC-SHA256 header on an incoming webhook.
 *
 * @param secret   - The WEBHOOK_SIGNING_SECRET shared with Dactyl (hex string).
 * @param rawBody  - The raw request body as a string or Buffer (before JSON parse).
 * @param signature - The value of the X-Dactyl-Signature header.
 * @returns true if the signature is valid, false otherwise.
 *
 * @example
 * ```ts
 * app.post('/webhook', (req, res) => {
 *   const valid = verifyDactylWebhook(
 *     process.env.WEBHOOK_SECRET!,
 *     req.rawBody,
 *     req.headers['x-dactyl-signature'] as string,
 *   );
 *   if (!valid) return res.status(401).send('Bad signature');
 *   // process event…
 * });
 * ```
 */
export function verifyDactylWebhook(
  secret: string,
  rawBody: string | Buffer,
  signature: string,
): boolean {
  try {
    const body = typeof rawBody === 'string' ? Buffer.from(rawBody) : rawBody;
    const expected = createHmac('sha256', secret).update(body).digest('hex');
    const expectedBuf = Buffer.from(expected, 'utf8');
    const signatureBuf = Buffer.from(signature, 'utf8');
    // Lengths must match for timingSafeEqual
    if (expectedBuf.length !== signatureBuf.length) {
      return false;
    }
    return timingSafeEqual(expectedBuf, signatureBuf);
  } catch {
    return false;
  }
}

File: sdk/python/dactyl_sdk/webhook.py
"""
Webhook signature verification helper for Dactyl webhooks.
"""
import hmac
import hashlib


def verify_dactyl_webhook(
    secret: str,
    raw_body: str | bytes,
    signature: str,
) -> bool:
    """
    Verify the X-Dactyl-Signature HMAC-SHA256 header on an incoming webhook.

    Args:
        secret:     The WEBHOOK_SIGNING_SECRET shared with Dactyl.
        raw_body:   The raw request body (str or bytes) before JSON parsing.
        signature:  The value of the X-Dactyl-Signature request header.

    Returns:
        True if the signature is valid, False otherwise.

    Example::

        from dactyl_sdk.webhook import verify_dactyl_webhook

        @app.post("/webhook")
        async def handle_webhook(request: Request):
            body = await request.body()
            sig = request.headers.get("x-dactyl-signature", "")
            if not verify_dactyl_webhook(WEBHOOK_SECRET, body, sig):
                raise HTTPException(status_code=401, detail="Bad signature")
            # process event…
    """
    try:
        body_bytes = raw_body.encode() if isinstance(raw_body, str) else raw_body
        expected = hmac.new(
            secret.encode(),
            body_bytes,
            hashlib.sha256,
        ).hexdigest()
        # Constant-time comparison
        return hmac.compare_digest(expected, signature)
    except Exception:
        return False
